File: multilibrary/src/main/java/com/zhongjh/albumcamerarecorder/camera/ui/camera/state/StateMode.java
package com.zhongjh.albumcamerarecorder.camera.ui.camera.state;

import com.zhongjh.albumcamerarecorder.camera.ui.camera.BaseCameraFragment;
import com.zhongjh.albumcamerarecorder.camera.ui.camera.manager.CameraPictureManager;
import com.zhongjh.albumcamerarecorder.camera.ui.camera.manager.CameraVideoManager;

/**
 * 状态模式
 *
 * @author zhongjh
 * @date 2021/11/25
 */
public abstract class StateMode implements IState {

    protected final String TAG = CameraStateManager.class.getSimpleName();

    CameraStateManager cameraStateManager;
    BaseCameraFragment<? extends CameraStateManager,
            ? extends CameraPictureManager,
            ? extends CameraVideoManager> cameraFragment;

    public BaseCameraFragment<? extends CameraStateManager,
            ? extends CameraPictureManager,
            ? extends CameraVideoManager> getCameraFragment() {
        return cameraFragment;
    }

    public void setCameraFragment(BaseCameraFragment<? extends CameraStateManager,
            ? extends CameraPictureManager,
            ? extends CameraVideoManager> cameraLayout) {
        this.cameraFragment = cameraLayout;
    }

    public CameraStateManager getCameraStateManagement() {
        return cameraStateManager;
    }

    public void setCameraStateManagement(CameraStateManager cameraStateManager) {
        this.cameraStateManager = cameraStateManager;
    }

    /**
     * @param cameraStateManager 可以让状态更改别的状态
     * @param cameraFragment       主要是多个状态围绕着cameraLayout进行相关处理
     */
    public StateMode(BaseCameraFragment<? extends CameraStateManager,
            ? extends CameraPictureManager,
            ? extends CameraVideoManager> cameraFragment, CameraStateManager cameraStateManager) {
        this.cameraFragment = cameraFragment;
        this.cameraStateManager = cameraStateManager;
    }

    @Override
    public void onActivityPause() {

    }

    @Override
    public Boolean onBackPressed() {
        return null;
    }

    @Override
    public void pvLayoutCommit() {

    }

    @Override
    public void pvLayoutCancel() {

    }

    @Override
    public void pauseRecord() {

    }

    @Override
    public void stopProgress() {

    }

    @Override
    public void onLongClickFinish() {

    }
}

File: multilibrary/src/main/java/com/zhongjh/multimedia/camera/ui/camera/state/type/impl/StateMode.kt
package com.zhongjh.multimedia.camera.ui.camera.state.type.impl

import com.zhongjh.multimedia.camera.ui.camera.BaseCameraFragment
import com.zhongjh.multimedia.camera.ui.camera.manager.CameraPictureViewManager
import com.zhongjh.multimedia.camera.ui.camera.manager.CameraVideoViewManager
import com.zhongjh.multimedia.camera.ui.camera.state.CameraStateManager
import java.lang.ref.WeakReference

/**
 * 状态模式
 *
 * @param cameraStateManager 可以让状态更改别的状态
 * @param cameraFragment       主要是多个状态围绕着cameraLayout进行相关处理
 *
 * @author zhongjh
 * @date 2021/11/25
 */
abstract class StateMode(
    var cameraFragment: BaseCameraFragment<out CameraStateManager, out CameraPictureViewManager, out CameraVideoViewManager>,
    private var cameraStateManager: CameraStateManager
) : IState {

    protected val tag: String = CameraStateManager::class.java.simpleName

    /**
     * 使用弱引用持有 Fragment 和 Manager
     */
    protected val fragmentRef = WeakReference(cameraFragment)
    protected val stateManagerRef = WeakReference(cameraStateManager)

    /**
     * 在 Fragment 销毁时调用，释放资源
     */
    open fun onDestroy() {
        fragmentRef.clear()
        stateManagerRef.clear()
    }

    override fun onActivityPause() {
    }

    override fun onBackPressed(): Boolean? {
        return null
    }

    override fun pvLayoutCommit() {
    }

    override fun pvLayoutCancel() {
    }

    override fun pauseRecord() {
    }

    override fun stopProgress() {
    }

    override fun onLongClickFinish() {
    }
}
